### ISSUE TITLE
(PS 5.1.14393) Passing commands via STDIN in Ruby does not seem to work

### ISSUE THREAD

**Issue by glennsarti** [PowerShell/PowerShell#2071]
Steps to reproduce
------------------
1. Install Ruby 2.3.0

2. Create a file in the root of the repository `test.rb` with the following content
```ruby
require 'open3'
require 'timeout'

cmd = 'powershell.exe -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command -'
str1 = <<-CODE


Write-Output "1"
Write-Output "2"
Write-Output "3"


CODE
# Helper
def drain_pipe(pipe, iterations = 10)
  output = []
  0.upto(iterations) do
    break if !is_readable?(pipe, 0.1)
    l = pipe.gets
    puts "#{Time.now} PIPE> #{l}"
    output << l
  end
  output
end
def self.is_readable?(stream, timeout = 0.5)
  read_ready = IO.select([stream], [], [], timeout)
  read_ready && stream == read_ready[0][0]
end


## Let's go
puts "Starting PS..."
@stdin, @stdout, @stderr, @ps_process = Open3.popen3(cmd)

puts "Enter to send STDIN"
blah = gets

puts "Sending STDIN..."
@stdin.puts(str1)
sleep(2)
output = drain_pipe(@stdout)

puts "output = #{output}"

puts "Enter to cleanup"
blah = gets

#  CLEANUP
puts "Cleaning up..."
@stdin.puts "\nexit\n"
@stdin.close
@stdout.close
@stderr.close

begin
Timeout.timeout(3) do
  puts "Awaiting PowerShell process termination..."
  @exit_status = @ps_process.value
end
rescue Timeout::Error
end

exit_msg = "PowerShell process exited: #{@exit_status}" if @exit_status
puts(exit_msg)
if @ps_process.alive?
  puts("Forcefully terminating PowerShell process.")
  Process.kill('KILL', @ps_process[:pid])
end
```

3. run 
```
ruby test.rb
```

4. Press enter when prompted

Expected behavior
-----------------
```
Starting PS...
Enter to send STDIN

Sending STDIN...
2016-08-25 16:10:24 -0700 PIPE> 1
2016-08-25 16:10:24 -0700 PIPE> 2
2016-08-25 16:10:24 -0700 PIPE> 3
output = ["1\n", "2\n", "3\n"]
Enter to cleanup

Cleaning up...
Awaiting PowerShell process termination...
PowerShell process exited: pid 2788 exit 0
```

Actual behavior
---------------
```
Starting PS...
Enter to send STDIN

Sending STDIN...
output = []
Enter to cleanup

Cleaning up...
Awaiting PowerShell process termination...

Forcefully terminating PowerShell process.
```

Environment data
----------------
Doesn't work
```
Windows 10 Enterprise


Name                           Value
----                           -----
PSVersion                      5.1.14393.82
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.82
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

Does Work
```
Server 2012 R2

Name                           Value
----                           -----
PSVersion                      5.0.10586.117
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.10586.117
CLRVersion                     4.0.30319.34209
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

```

Workaround
----------------
Modify the initial powershell command from
`cmd = 'powershell.exe -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command -'`

to

`cmd = 'powershell.exe -Version 2.0 -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command -'`

This works on the Windows 10 system (PS 5.1)

**Comment by glennsarti** [comment]
May be linked to https://github.com/PowerShell/PowerShell/issues/1794

**Comment by glennsarti** [comment]
Further investigation.

I installed IONinja and sniffed the named pipes being used between ruby and powershell

Working test
----
In this case I used the `-Version 2.0` flag to the initial powershell call
This is the logfile from IO Ninja
```
09:39:30 +01:51.987   File #9: Server file opened: \ruby00000000000013f4-00000000
09:39:30 +01:51.988   File #10: Client file opened: \ruby00000000000013f4-00000000
09:39:30 +01:52.003   File #9: 
09:39:30 +01:52.002 > 0000  41 63 74 69 76 65 20 63 6f 64 65 20 70 61 67 65  Active code page
                    > 0010  3a 20 38 35 30 0d 0a                             : 850..
09:39:30 +01:52.003   File #10: 
09:39:30 +01:52.002 < 0000  41 63 74 69 76 65 20 63 6f 64 65 20 70 61 67 65  Active code page
                    < 0010  3a 20 38 35 30 0d 0a                             : 850..
09:39:30 +01:52.005   File closed
09:39:30 +01:52.005   File #9: File closed
09:39:31 +01:52.649   File #11: Server file opened: \ruby0000000000001d98-00000000
09:39:31 +01:52.649   File #12: Client file opened: \ruby0000000000001d98-00000000
09:39:31 +01:52.649   File #13: Server file opened: \ruby0000000000001d98-00000001
09:39:31 +01:52.649   File #14: Client file opened: \ruby0000000000001d98-00000001
09:39:31 +01:52.649   File #15: Server file opened: \ruby0000000000001d98-00000002
09:39:31 +01:52.649   File #16: Client file opened: \ruby0000000000001d98-00000002
09:39:35 +01:56.681   File #11: 
09:39:35 +01:56.681 > 0000  0a 0a 57 72 69 74 65 2d 4f 75 74 70 75 74 20 22  ..Write-Output "
                    > 0010  31 22 0a 57 72 69 74 65 2d 4f 75 74 70 75 74 20  1".Write-Output 
                    > 0020  22 32 22 0a 57 72 69 74 65 2d 4f 75 74 70 75 74  "2".Write-Output
                    > 0030  20 22 33 22 0a 0a 0a                              "3"...
09:39:35 +01:56.681   File #12: 
09:39:35 +01:56.681 < 0000  0a 0a 57 72 69 74 65 2d 4f 75 74 70 75 74 20 22  ..Write-Output "
                    < 0010  31 22 0a 57 72 69 74 65 2d 4f 75 74 70 75 74 20  1".Write-Output 
                    < 0020  22 32 22 0a 57 72 69 74 65 2d 4f 75 74 70 75 74  "2".Write-Output
                    < 0030  20 22 33 22 0a 0a 0a                              "3"...
09:39:35 +01:56.723   File #14: 
09:39:35 +01:56.729 < 0000  31 0d 0a 32 0d 0a 33 0d 0a                       1..2..3..
09:39:37 +01:58.681   File #13: 
09:39:37 +01:58.681 > 0000  31 0d 0a 32 0d 0a 33 0d 0a                       1..2..3..
09:40:26 +02:47.551   File #11: 
09:40:26 +02:47.551 > 0000  0a 65 78 69 74 0a                                .exit.
09:40:26 +02:47.551   File #12: 
09:40:26 +02:47.551 < 0000  0a 65 78 69 74 0a                                .exit.
09:40:26 +02:47.551   File closed
09:40:26 +02:47.551   File #13: File closed
09:40:26 +02:47.551   File #15: File closed
09:40:26 +02:47.562   File #11: File closed
09:40:26 +02:47.562   File #14: File closed
09:40:26 +02:47.562   File #16: File closed
```

Note that `File #12` is the outbound pipe from ruby into powershell STDIN and `File #11` is in the inbound pipe into Powershell STDIN from ruby.  See that the content from `File #12` equals the content to `File #11`.

They appear out of order in the log because the timestamps are the same

You can see the inverse behaviour in `File #14` and `File #13` which his STDOUT pipe from powershell to ruby.  These are it the correct oder as you can see by the timestamps


Failing Test
--------
This time I used the powershell command without a version specifier i.e. PS 5.1
Log file
```
09:37:18 -00:20.199   Session started
09:37:18 -00:20.199   Capture started with filter *ruby0000000000002a58-00000001*
09:37:37 -00:00.720   Capture stopped
09:37:38 +00:00.000   Session started
09:37:38 +00:00.000   Capture started with filter *ruby00000000*
09:37:57 +00:19.019   File #1: Server file opened: \ruby000000000000334c-00000000
09:37:57 +00:19.019   File #2: Client file opened: \ruby000000000000334c-00000000
09:37:57 +00:19.032   File #1: 
09:37:57 +00:19.032 > 0000  41 63 74 69 76 65 20 63 6f 64 65 20 70 61 67 65  Active code page
                    > 0010  3a 20 38 35 30 0d 0a                             : 850..
09:37:57 +00:19.032   File #2: 
09:37:57 +00:19.032 < 0000  41 63 74 69 76 65 20 63 6f 64 65 20 70 61 67 65  Active code page
                    < 0010  3a 20 38 35 30 0d 0a                             : 850..
09:37:57 +00:19.034   File closed
09:37:57 +00:19.035   File #1: File closed
09:37:58 +00:19.582   File #3: Server file opened: \ruby0000000000001f9c-00000000
09:37:58 +00:19.582   File #4: Client file opened: \ruby0000000000001f9c-00000000
09:37:58 +00:19.582   File #5: Server file opened: \ruby0000000000001f9c-00000001
09:37:58 +00:19.582   File #6: Client file opened: \ruby0000000000001f9c-00000001
09:37:58 +00:19.582   File #7: Server file opened: \ruby0000000000001f9c-00000002
09:37:58 +00:19.582   File #8: Client file opened: \ruby0000000000001f9c-00000002
09:38:09 +00:31.379   File #4: 
09:38:09 +00:31.379 < 0000  0a 0a 57 72 69 74 65 2d 4f 75 74 70 75 74 20 22  ..Write-Output "
                    < 0010  31 22 0a 57 72 69 74 65 2d 4f 75 74 70 75 74 20  1".Write-Output 
                    < 0020  22 32 22 0a 57 72 69 74 65 2d 4f 75 74 70 75 74  "2".Write-Output
09:38:28 +00:50.172 < 0030  20 22 33 22 0a 0a 0a 0a 65 78 69 74 0a            "3"....exit.
09:38:28 +00:50.172   File closed
09:38:28 +00:50.172   File #5: File closed
09:38:28 +00:50.172   File #7: File closed
09:38:31 +00:53.176   File #3: File closed
09:38:31 +00:53.176   File #6: File closed
09:38:31 +00:53.176   File #8: File closed
```
Note that `File #4` is the outbound pipe from ruby into powershell STDIN and `File #3 ` is in the inbound pipe into Powershell STDIN from ruby.

See that the data is put into the pipe but it never gets read out at the powershell pipe end.  This suggests there's something up with the pipe handling.

**Comment by lzybkr** [comment]
I updated the repro making it slightly simpler (cloning puppetlabs-powershell wasn't necessary, just needed to add `require 'timeout'` to the script and it runs just fine with just `ruby` installed.

When I run the modified repro, it does hang when running the in-box `powershell.exe`, but exits correctly with my fix to #2090.

That said, the output is still missing. I verified that PowerShell is receiving the command and executing them, but the ruby script stopped reading from `stdout` before the commands ran. I increased the timeout and verified the output was received.

So I think this issue is the same as #2090, but it would be great if you can confirm.
